Add performOnce to enqueue a Job with no retries

pbrisbin opened the pull request:
**Breaking changes**: `performAt` and `performIn` now accept the varying
argument (the `time` or `diff`) first instead of last. This made their
implementations in terms of an internal `performWith` function more
natural.

**Motivating need**: we want to modify the retries attribute of a Job
to, well, prevent it from retrying on errors. This is useful for Jobs
that include some *at most once* action, such as sending an Email.

**Alternatives explored**:

1. Most minimal

   We actually have everything we need available in the public interface
   to do this already, *except* the actual fields of `Job`. So the most
   minimal change we could make to support this need would be to just
   change the export from `Job` to `Job(..)`. I decided to accept some
   of the end-user burden within the library and explore providing
   something for this use-case directly.

2. Most backwards-compatible

   I could just add `performOnce`, with no other changes.

   I decided that taking that approach would exacerbate the (admittedly
   minor) duplication issues between `performAt`, `performIn`, and
   `performOnce`, which (IMO) really ought to be variations on a theme,
   i.e. implemented in terms of a shared `performWith(M)`.

   Doing that most naturally meant moving the varying arguments in
   `performAt` and `performIn` to first position instead of last -- a
   breaking change. I chose to accept this given the library is so
   young.

3. Most flexible

   I could've *only* implemented `performWith(M)` and exported that,
   leaving it again up to the end-user to implement (something like)
   `performOnce` using it.

   This would require exporting `Job(..)` so the end-user could set
   `jobRetry`. I think if I were willing to expose more of our internals
   like that, I would've just went with Alternative (1) to begin with.

cdparks commented on library/Faktory/Job.hs:
```diff
@@ -35,26 +36,50 @@ data Job arg = Job
+  -> IO JobId
+performWithM f client queue arg = do
+  job <- f =<< newJob queue arg
+  jobJid job <$ pushJob client job
```
I'm fine with this, but if you really wanna go wild, you can use the monoidal options pattern:

```haskell
perform mempty "queue" SomeJob
perform once "queue" SomeJob
perform (once <> at someTime) "queue" SomeJob
perform (once <> in_ 10) "queue" SomeJob
perform (withRetries 3 <> in_ 10) "queue" SomeJob
```

i.e.
```haskell
-- Simplified, cuz i did this real fast
type Queue = String

data Job a = Job
  { jobRetry :: !Int
  , jobQueue :: !String
  , jobAt :: !(Maybe UTCTime)
  , jobArgs :: !a
  } deriving (Eq, Show, Ord)

data JobUpdate
  = SetRetries Int
  | SetAt UTCTime
  | SetIn NominalDiffTime
  deriving (Eq, Show, Ord)

newtype JobOptions = JobOptions [JobUpdate]
  deriving newtype (Eq, Show, Ord, Semigroup, Monoid)

perform :: JobOptions -> Queue -> a -> IO (Job a)
perform options q a = applyOptions options $ newJob q a

newJob :: Queue -> a -> Job a
newJob q a = Job
  { jobRetry = 25
  , jobQueue = q
  , jobAt = Nothing
  , jobArgs = a
  }

applyOptions :: JobOptions -> Job a -> IO (Job a)
applyOptions (JobOptions patches) = go patches
 where
  go [] job = pure job
  go (x:xs) job =
    case x of
      SetRetries n ->
        go xs $ job {jobRetry = n}
      SetAt t ->
        go xs $ job {jobAt = Just t}
      SetIn i -> do
        t <- getCurrentTime
        go xs $ job {jobAt = Just $ addUTCTime i t}

withRetries :: Int -> JobOptions
withRetries n = JobOptions [SetRetries n]

once :: JobOptions
once = withRetries 0

at :: UTCTime -> JobOptions
at t = JobOptions [SetAt t]

in_ :: NominalDiffTime -> JobOptions
in_ i = JobOptions [SetIn i]
```

Note that this doesn't prevent you from setting the same options more than once, though you can do that with fancier types. Probably not worth the trouble

pbrisbin replied:
I like this. I like that it avoids `performWithM`. And I like that it's just one `perform` function, having to supply `mempty` in the simplest case seems reasonable to me.

cdparks commented:
Re: breaking changes, thanks for going through your thought process, but I'm fine with them as long as we're not on hackage or stackage yet. I'm fairly sure we're the only users of this library. Feel free to bump version though

pbrisbin commented:
@cdparks I went with the Monoid suggestion, very nice. There are some minor differences:

- I tried to keep the naming consistent with the Faktory protocol, so I use `retry` instead of `retries` a lot
- I think that's it, but I can't remember :)

I also went a little further:

- I moved the `Queue` argument into `JobOptions` with another `Set` case
- I added support for setting `Jobtype` the same way
- I move the `Queue` argument on the _worker_ side into the `Settings` value we already
- And added support for `FAKTORY_QUEUE` in the env-based settings configuration

I think this simplifies/unifies well, but the worker-side changes might be a bit drastic. Let me know what you think.

(NOTE: I see you Restyled; I'll apply your changes at the end after this is approved and history is cleaned up.)

pbrisbin commented:
Nice, thanks @cdparks! Any pushes from now should just be getting CI green then cleaning up history; feel free to ignore.
